### ISSUE TITLE
Docker-compose container hangs on exec, version 18.06.1

### ISSUE THREAD

**Issue by SpencerMalone** [moby/moby#37764]
**Description**

Semi regularly (once every other day on ~1/10 hosts?) our jenkins build hosts have docker containers hang. The containers that hang are created and exec'd into with compose, but I'm unsure if that's the cause, or if docker/moby itself is the root cause. ~The `docker-runc` process goes defunct, as does all the commands related to the `exec`~ (I've since seen cases where this locks up with the same stack trace, but does not have defunct commands), but not the original entrypoint or `docker-containerd-shim`. You cannot inspect, stop, or kill the container until someone manually `kill -9`'s the root `docker-containerd-shim` proc. 

**Steps to reproduce the issue:**
Unfortunately I've not found a reliable way to reproduce this. Our script is doing the following:
`docker-compose up -d`
`docker-compose exec -T app bash -l -c "/opt/scripts/unit-tests.sh"` (which runs some unit tests)
and then it hangs on occasion. I can't quite tell if it's hanging on this, or hanging on a later `docker-compose down`, but the golang dumps I got make me think it's this.

These compose files do specify `cpuset: 1-${MAXCPU}` so that they leave 1 CPU unused at all times. Other than that, they are fairly bog standard nginx+phpfpm app containers with a percona sidecar container used to run some integration tests.

**Describe the results you received:**
Any command that interacts with the container hangs until restarted

**Describe the results you expected:**
Graceful ending of commands

**Additional information you deem important (e.g. issue happens only occasionally):**

https://gist.github.com/SpencerMalone/69edeb9cc65289d402ea36b11d8ac3e4 <- Gist of the golang dump

Of particular interest to me was... 
```
goroutine 163 [semacquire, 192 minutes]:
sync.runtime_Semacquire(0xc422a0b6b4)
	/usr/local/go/src/runtime/sema.go:56 +0x3b
sync.(*RWMutex).RLock(0xc422a0b6a8)
	/usr/local/go/src/sync/rwmutex.go:50 +0x4b
github.com/docker/docker/daemon/exec.(*Store).List(0xc422a0b6a0, 0xc424576240, 0x2, 0x2)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/exec/exec.go:140 +0x4d
github.com/docker/docker/daemon.(*Daemon).containerExecIds(0xc42041ac40, 0xc4207e9f50)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/exec.go:319 +0xe9
github.com/docker/docker/daemon.(*Daemon).execCommandGC(0xc42041ac40)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/exec.go:296 +0x17f
created by github.com/docker/docker/daemon.NewDaemon
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/daemon.go:878 +0x218d
```

`ps` grepped to the docker-containerd-shim proc id:
``` 
ps -ef | grep 41917
root     20663 41917  0 Sep04 ?        00:00:00 [php] <defunct>
root     22589 41917  0 Sep04 ?        00:00:00 [php] <defunct>
root     29564 41917  0 Sep04 ?        00:00:00 [php] <defunct>
root     30596 41917  0 Sep04 ?        00:00:00 [php] <defunct>
root     40418 41917  0 Sep04 ?        00:00:00 [php] <defunct>
root     41917 18771  0 Sep04 ?        00:00:01 docker-containerd-shim -namespace moby -workdir /var/lib/docker/containerd/daemon/io.containerd.runtime.v1.linux/moby/4361cd747382b352a90a88bebce8727bef3fff452288799dc4ed6466e3c5c8ac -address /var/run/docker/containerd/docker-containerd.sock -containerd-binary /usr/bin/docker-containerd -runtime-root /var/run/docker/runtime-runc
root     41935 41917  0 Sep04 ?        00:00:00 sh -l -c /usr/sbin/php-fpm && nginx
root     42999 41917  0 Sep04 ?        00:00:00 [docker-runc] <defunct>
root     45699  9344  0 01:36 pts/0    00:00:00 grep --color=auto 41917
```

**Output of `docker version`:**

```
docker -v
Docker version 18.06.1-ce, build e68fc7a
```

**Output of `docker info`:**

```
docker info
Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 113
Server Version: 18.06.1-ce
Storage Driver: devicemapper
 Pool Name: docker-8:3-806104106-pool
 Pool Blocksize: 65.54kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Udev Sync Supported: true
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Data Space Used: 18.35GB
 Data Space Total: 107.4GB
 Data Space Available: 89.02GB
 Metadata Space Used: 29.2MB
 Metadata Space Total: 2.147GB
 Metadata Space Available: 2.118GB
 Thin Pool Minimum Free Space: 10.74GB
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-09-28)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.110-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 48
Total Memory: 62.79GiB
Name: batch09.jenkins2.prod.atl01.rsglab.com
ID: NVIS:626W:JHIK:OOUE:4WZA:HHT4:6UFS:JEVU:L2OJ:PIIR:2XK7:62KQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: devicemapper: usage of loopback devices is strongly discouraged for production use.
         Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```
Please note that we have some hosts running a similar config with overlay and overlay2 that have also had lockups, but we've not dumped the threads on those to confirm it's locking at the same point. Next time they repeat, I'll be sure to do that.

**Additional environment details (AWS, VirtualBox, physical, etc.):**
CentOS Linux release 7.3.1611 (Core)
Physical box that may have cruft around from prior docker versions. We've done our best to ensure we don't leave that around during upgrades, but stuff sneaks by.

**Comment by SpencerMalone** [comment]
Had this happen on another host, this one running overlay2 (just noticed it has a slight misconfiguration, but the same error cropped up).

Docker info:
```
Containers: 9
 Running: 8
 Paused: 0
 Stopped: 1
Images: 14
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: false
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.110-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 48
Total Memory: 62.79GiB
Name: batch05.jenkins2.prod.atl01.rsglab.com
ID: NOOY:F7NP:L43C:VPTC:A4QQ:H5L2:GVH2:ZZGX:MLAG:EZDV:EO2C:IQO2
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: overlay2: the backing xfs filesystem is formatted without d_type support, which leads to incorrect behavior.
         Reformat the filesystem with ftype=1 to enable d_type support.
         Running without d_type support will not be supported in future releases.
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

Gist of golang log dump using SIGUSR1 signal: https://gist.github.com/SpencerMalone/99226defea3c55d7926d4f5f4f7f4c2d

Same centos release

**Comment by SpencerMalone** [comment]
Note, I am using `docker-compose version 1.22.0, build f46880fe` in case that matters

**Comment by jrosdahl** [comment]
I'm seeing this or a very similar problem as well. I'm not using docker-compose though.

* I use Docker CE 18.03 on CentOS 7.5.
* I start a container that executes "sleep forever" via init.
* I exec build commands in the container.
* When killing the container before the build has finished, the associated docker-containerd-shim process sometimes hangs and the container processes are marked as defunct like this:

```
 PPID   PID  PGID   SID TTY      TPGID STAT   UID   TIME COMMAND
    1 16811 16811 16811 ?           -1 Ssl      0 243:56 /usr/bin/dockerd
16811 16825 16825 16825 ?           -1 Ssl      0 181:30  \_ docker-containerd --config /var/run/docker/containerd/containerd.toml
16825 20688 20688 16825 ?           -1 Sl       0   0:00      \_ docker-containerd-shim -namespace moby -workdir /extra/docker-data/containerd/daemon/io.containerd.runtime.v1.linux/
moby/78f02a07bb6e41068bd3a72565e95e6df2e476cd00a73a2d0210976796551e96 -address /var/run/docker/containerd/docker-containerd.sock -containerd-binary /usr/bin/docker-containerd -runti
me-root /var/run/docker/runtime-runc
20688 20704 20704 20704 ?           -1 Ds   65000   0:00          \_ [init]
20688 28607 28469 28469 ?           -1 Z    65000   0:00          \_ [make] <defunct>
20688 28725 28469 28469 ?           -1 Z    65000   0:00          \_ [sh] <defunct>
[...]
```

Strace says that docker-containerd-shim waits on a futex:

```
# strace -p 20688
strace: Process 20688 attached
futex(0x821bf8, FUTEX_WAIT, 0, NULL
```

Any further interaction with the container hangs. Sending SIGKILL to said docker-containerd-shim process makes the container tear down as expected.

containerd stack trace: https://gist.github.com/jrosdahl/1467d7cef99592383a54f12859e5b031

**Comment by SpencerMalone** [comment]
I tried removing my `docker-compose exec` calls last week and can confirm that it's not tied to that. I also have this same behavior of being unable to cleanup until sending the SIGKILL, and that definitely knocks it offline.  Have you found a way to reliably reproduce this? I can loop my task all day and it eventually happens, but I've not found a "quick" way to trigger this.

Also, any chance it's something about the images? We're using some home baked centos images built from scratch, so I figure it's worth seeing if you are doing something similar.

**Comment by jrosdahl** [comment]
> Have you found a way to reliably reproduce this?

My scenario is a CI system, and the problem is fairly easy to reproduce for me, yes. Perhaps one in three attempts triggers the issue, where "an attempt" means starting a build and killing it while running. I haven't had time to try to minimize the context to create a standalone reproduction recipe, unfortunately.

> Also, any chance it's something about the images? We're using some home baked centos images built from scratch, so I figure it's worth seeing if you are doing something similar.

I'm using a custom image based on latest the centos7 image.

**Comment by cpuguy83** [comment]
Can you grab the stack trace from containerd?

```
docker-containerd-ctr pprof --debug-socket /run/docker/containerd/docker-containerd-debug.sock goroutines
```

I believe if you are on 18.09 this debug socket path would be `containerd-debug.sock` (no `docker-`).

**Comment by SpencerMalone** [comment]
Hrm, sorry for the slow response, we optimized for recovery (by deleting the stuck container shims procs when we find them), and this issue became transparent to our users. Looking back on our logs, this stopped happening? I'm trying to correlate the decrease in this occurring with something, I know we did a big round of OS updates for specter stuff, maybe that kernel version bump saved us?

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
